code/modules/ninja/suit/gloves.dm

```diff
@@ -28,303 +36,37 @@
-					PN.load += drained
-					if(drained < drain)//if no power on net, drain apcs
-						for(var/obj/machinery/power/terminal/T in PN.nodes)
-							if(istype(T.master, /obj/machinery/power/apc))
-								var/obj/machinery/power/apc/AP = T.master
-								if(AP.operating && AP.cell && AP.cell.charge>0)
-									AP.cell.charge = max(0, AP.cell.charge - 5)
-									drained += 5
-				else	break
-				S.cell.charge += drained
-				if(S.cell.charge>S.cell.maxcharge)
-					totaldrain += (drained-(S.cell.charge-S.cell.maxcharge))
-					S.cell.charge = S.cell.maxcharge
-					maxcapacity = 1
-				else
-					totaldrain += drained
-				S.spark_system.start()
-				if(drained==0)	break
-			U << "<span class='notice'>Gained <B>[totaldrain]</B> energy from the power network.</span>"
-
-		if("MECHA")
-			var/obj/mecha/A = target
-			A.occupant_message("<span class='danger'>Warning: Unauthorized access through sub-route 4, block H, detected.</span>")
-			if(A.get_charge())
-				while(G.candrain&&A.cell.charge>0&&!maxcapacity)
-					drain = rand(G.mindrain,G.maxdrain)
-					if(A.cell.charge<drain)
-						drain = A.cell.charge
-					if(S.cell.charge+drain>S.cell.maxcharge)
-						drain = S.cell.maxcharge-S.cell.charge
-						maxcapacity = 1
-					if (do_after(U,10))
-						A.spark_system.start()
-						playsound(A.loc, "sparks", 50, 1)
-						A.cell.use(drain)
-						S.cell.charge+=drain
-						totaldrain+=drain
-					else	break
-				U << "<span class='notice'>Gained <B>[totaldrain]</B> energy from [src].</span>"
-			else
-				U << "<span class='danger'>The exosuit's battery has run dry. You must find another source of power.</span>"
-
-		if("CYBORG")
-			var/mob/living/silicon/robot/A = target
-			A << "<span class='danger'>Warning: Unauthorized access through sub-route 12, block C, detected.</span>"
-			G.draining = 1
-			if(A.cell&&A.cell.charge)
-				while(G.candrain&&A.cell.charge>0&&!maxcapacity)
-					drain = rand(G.mindrain,G.maxdrain)
-					if(A.cell.charge<drain)
-						drain = A.cell.charge
-					if(S.cell.charge+drain>S.cell.maxcharge)
-						drain = S.cell.maxcharge-S.cell.charge
-						maxcapacity = 1
-					if (do_after(U,10))
-						A.spark_system.start()
-						playsound(A.loc, "sparks", 50, 1)
-						A.cell.charge-=drain
-						S.cell.charge+=drain
-						totaldrain+=drain
-					else	break
-				U << "<span class='notice'>Gained <B>[totaldrain]</B> energy from [A].</span>"
-			else
-				U << "<span class='danger'>Their battery has run dry of power. You must find another source.</span>"
-
-		else//Else nothing :<
+	draining = 1
+	var/drained = A.ninjadrain_act(suit,H,src)
+	draining = 0
 
-	G.draining = 0
+	if(isnum(drained)) //Numerical values of drained handle their feedback here, Alpha values handle it themselves (Research hacking)
+		if(drained)
+			H << "<span class='notice'>Gained <B>[drained]</B> energy from \the [A].</span>"
+		else
+			H << "<span class='danger'>\this [A] has run dry of power, you must find another source!</span>"
```

Comment:
```"\This is not a valid source of power"```? Covers both invalid objects and objects without power in the same message. @phil235